File: business-logic/classic-scanlog-core/tests/semantic_conformance/yaml_source_values.rs
//! Generic YAML source path and display semantics from the public core enum.
use super::{RunnerResult, invalid, text};
use classic_config_core::YamlSource;
use serde_json::{Value, json};
use std::collections::HashSet;

/// Resolve fixture-selected public variants and retain the complete source matrix.
pub(super) fn execute(fixture: &Value) -> RunnerResult<Value> {
    let game = text(&fixture["game"])?;
    let mut sources = Vec::new();
    let mut rows = Vec::new();
    for raw in fixture["sources"]
        .as_array()
        .ok_or_else(|| invalid("sources must be an array"))?
    {
        let name = text(raw)?;
        let source = match name.as_str() {
            "MAIN" => YamlSource::Main,
            "IGNORE" => YamlSource::Ignore,
            "GAME" => YamlSource::Game,
            "GAME_LOCAL" => YamlSource::GameLocal,
            "TEST" => YamlSource::Test,
            _ => return Err(invalid("unsupported source variant").into()),
        };
        rows.push(json!({"id": name, "path": source.path(&game).to_string_lossy().replace('\\', "/"), "name": source.display_name(), "gameName": source.display_name_with_game(&game)}));
        sources.push(source);
    }
    Ok(
        json!({"sources": rows, "equal": sources.first() == Some(&YamlSource::Main), "different": sources.first() != sources.get(1), "distinct": sources.iter().collect::<HashSet<_>>().len()}),
    )
}

File: tools/binding_compliance/conformance/families/yaml_source_values.py
"""Generic YAML source paths and display labels from real public adapters."""

from collections.abc import Mapping

from ..coverage import CoveragePredicate, FamilyCoveragePolicy


def matches_sources(observation: Mapping) -> bool:
    """Require a complete ordered source matrix and observed identity distinctions."""
    rows = observation.get("sources")
    return (
        set(observation) == {"sources", "equal", "different", "distinct"}
        and observation["equal"] is True
        and observation["different"] is True
        and observation["distinct"] == 5
        and isinstance(rows, list)
        and [row.get("id") for row in rows if isinstance(row, Mapping)]
        == ["MAIN", "IGNORE", "GAME", "GAME_LOCAL", "TEST"]
        and all(
            set(row) == {"id", "path", "name", "gameName"}
            and all(isinstance(value, str) and value for value in row.values())
            for row in rows
        )
        and rows[2]["path"] == "CLASSIC Data/databases/CLASSIC Fallout4.yaml"
        and rows[3]["path"] == "CLASSIC Data/CLASSIC Fallout4VR Local.yaml"
    )


YAML_SOURCE_VALUES_COVERAGE_POLICY = FamilyCoveragePolicy(
    "yaml-source-values",
    (
        CoveragePredicate(
            "yaml-source-values.matrix",
            "yaml-source-values.matrix",
            "yaml-source-values.matrix",
            "values",
            ("YamlSource",),
            matches_sources,
            runtime_operations=(
                None,
                "path",
                "display_name",
                "display_name_with_game",
                "__eq__",
                "__hash__",
                "__repr__",
                "__str__",
                "getYamlSourcePath",
                "getYamlSourceDisplayName",
                "getYamlSourceDisplayNameWithGame",
            ),
        ),
    ),
)
